### ISSUE TITLE
BUG pickling formula information 

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#1263]
AFAIU:
The formula information is attached to the dataframe orig_exog. Pickling and unpickling the results and model instance looses the information.
Predict which relies on this information if `transform=True`, the default, is then broken and raises an exception

see http://stackoverflow.com/questions/20724919/pandas-dataframe-attributeerror-dataframe-object-has-no-attribute-design-inf/20726340

```
    p = result.predict(newdf)
  File "C:\Python27\lib\site-packages\statsmodels\base\model.py", line 878, in predict
    exog = dmatrix(self.model.data.orig_exog.design_info.builder,
  File "C:\Python27\lib\site-packages\pandas\core\frame.py", line 2088, in __getattr__
    (type(self).__name__, name))
AttributeError: 'DataFrame' object has no attribute 'design_info'
```

test case

```
import cPickle
import pandas as pd
import numpy as np
import statsmodels.formula.api as sm

df = pd.DataFrame({"A": [10,20,30,324,2353], "B": [20, 30, 10, 1, 2332], "C": [0, -30, 120, 11, 2]})

result = sm.ols(formula="A ~ B + C", data=df).fit()
print result.summary()

test1 = result.predict(df) #works

f_myfile = open('resultobject', "wb")
cPickle.dump(result, f_myfile, 2)
f_myfile.close()
print("Result Object Saved")


f_myfile = open('resultobject', "rb")
model = cPickle.load(f_myfile)

test2 = model.predict(df) #produces error
```

**Comment by jseabold** [comment]
I'm thinking we should just raise an informative error here. I don't see an easy way to pickle objects created with a formula.

@njsmith

**Comment by josef-pkt** [comment]
We won't get a fix for 0.7, since there is not much we can do.

But this is a bug and issue should remain open (indefinitely).

I think we might be able to get it partially, if the `eval` environment is not needed.

**Comment by jseabold** [comment]
I might have a workaround.

**Comment by jseabold** [comment]
Thought this doesn't work right now even if you don't use a formula AFAICT

**Comment by jseabold** [comment]
Strike that. I broke it.

**Comment by josef-pkt** [comment]
We have unit tests for pickling in `base` for the case when endog and exog are ndarrays, but currently no other case is tested and "officially" supported.

**Comment by chrimobs** [comment]
Sorry for bothering, but does this mean the bug is to be considered resolved? If I try to load a pickled result that uses a categorical variable and get NameError that *$column_name* of the categorical variable is not defined.

```
#!/usr/bin/python

import cPickle as cpick
import pandas as pan
import statsmodels.formula.api as smapi

data_df = pan.DataFrame({'p':[20,21,22,23,24,26,28,29,30,31], 'c':[0., 1.,
    1., 3., 3., 4., 4., 4.95, 6.31, 7.], 'a':[0,0,0,1,1,1,1,1,1,1]})

pred_df = pan.DataFrame({'c':[1., 2., 2., 4., 4., 5., 5., 6., 7., 8.],
    'a':[0,0,0,0,1,1,1,1,1,1]})

res = smapi.ols(formula='p ~ c:C(a)', data = data_df).fit()

print 'predict with orig data'
print res.predict(data_df)

with open('pickled_res', 'wb') as f:
    cpick.dump(res, f, 2)

with open('pickled_res', 'rb') as f:
    model = cpick.load(f)

print 'predict with unpickled model and new data'
print model.predict(pred_df)
```
If, however, the formula is rather `p ~ c` or `p ~ c+a` unpickling and `predict()' works fine.

**Comment by josef-pkt** [comment]
I don't see any unit tests for this for categorical explanatory variables. So, it's possible that there are still problems with some kinds of formulas.

**Comment by josef-pkt** [comment]
I put temporarily prio-high label on this. (I doubt I will have time to dig into this very soon.)

And thanks for reporting with a complete example. 

adjusted to run with python 3 (which I'm currently using for development)
also we have load and save methods that can be used directly

```
#!/usr/bin/python

#import cPickle as cpick
from statsmodels.compat import cPickle as cpick
import pandas as pan
import statsmodels.formula.api as smapi

data_df = pan.DataFrame({'p':[20,21,22,23,24,26,28,29,30,31], 'c':[0., 1.,
    1., 3., 3., 4., 4., 4.95, 6.31, 7.], 'a':[0,0,0,1,1,1,1,1,1,1]})

pred_df = pan.DataFrame({'c':[1., 2., 2., 4., 4., 5., 5., 6., 7., 8.],
    'a':[0,0,0,0,1,1,1,1,1,1]})

#res = smapi.ols(formula='p ~ c:C(a)', data = data_df).fit()
res = smapi.ols(formula='p ~ c + a', data = data_df).fit()

print('predict with orig data')
print(res.predict(data_df))

##with open('pickled_res', 'wb') as f:
##    cpick.dump(res, f, 2)
##
##with open('pickled_res', 'rb') as f:
##    model = cpick.load(f)

res.save('pickled_res')
model = res.load('pickled_res')

print('predict with unpickled model and new data')
print(model.predict(pred_df))
```

**Comment by njsmith** [comment]
FYI @chrish42 got a start towards making patsy builders pickleable (which is the only real solution here) at the pycon sprints -- see pydata/patsy#66.

**Comment by chrimobs** [comment]
As a motivation: Would it make any difference knowing that my dissertation may depend on this (proper unpickling when the model (the `RegressionResult`) contains transformed variables)? :) 
*Maybe* I could go without categorical variables, but the models do not perform well without them. I'm pondering on workarounds in my evaluation suite, but having that unpickling/prediction at hand would be **very** nice.

**Comment by josef-pkt** [comment]
@chrimobs I don't think having to drop categorical variables is a reasonable option.
What is your use case for pickling? Maybe there is an easier way to achieve the same task, at least until this is fixed.
Do you need the transformation in the post-estimation results, like predict, or can you transform all your data at once?

**Comment by josef-pkt** [comment]
stepping up from the exception with pdb, the problem seems to be that we give patsy the transformed data and not the original data. 
I think we don't keep the original dataframe around. In the example, the original data is in `frame` during pickle load.

```
(Pdb) data
    p  Intercept  c:C(a)[0]  c:C(a)[1]
0  20          1          0       0.00
1  21          1          1       0.00
2  22          1          1       0.00
3  23          1          0       3.00
4  24          1          0       3.00
5  26          1          0       4.00
6  28          1          0       4.00
7  29          1          0       4.95
8  30          1          0       6.31
9  31          1          0       7.00
```

From this it looks like no transformation would work after unpickling. ???

**Comment by josef-pkt** [comment]
Ok, that works for the example, use `frame` if available, else use current behavior

```
            try:
                data = d['frame']
            except KeyError:
                data = d['orig_endog'].join(d['orig_exog'])
```

The pickle file might be pretty large if we save several copies of the data in this case.

Unit tests for low memory pickling, after removing data that is not needed for prediction, are only for use with ndarrays.

**Comment by chrimobs** [comment]
My problem is that unpickling already fails if the original data was transformed to fit the model. This is what I get in such a case:
```
Traceback (most recent call last):
  File "./ru_wl_eval.py", line 26, in <module>
    res = smlm.RegressionResults.load(res_f)
  File "/usr/local/lib/python2.7/dist-packages/statsmodels-0.7.0-py2.7-linux-i686.egg/statsmodels/base/model.py", line 1675, in load
    return load_pickle(fname)
  File "/usr/local/lib/python2.7/dist-packages/statsmodels-0.7.0-py2.7-linux-i686.egg/statsmodels/iolib/smpickle.py", line 41, in load_pickle
    return cPickle.load(fin)
  File "/usr/local/lib/python2.7/dist-packages/statsmodels-0.7.0-py2.7-linux-i686.egg/statsmodels/base/data.py", line 98, in __setstate__
    return_type='dataframe')
  File "build/bdist.linux-i686/egg/patsy/highlevel.py", line 297, in dmatrices
  File "build/bdist.linux-i686/egg/patsy/highlevel.py", line 152, in _do_highlevel_design
  File "build/bdist.linux-i686/egg/patsy/highlevel.py", line 57, in _try_incr_builders
  File "build/bdist.linux-i686/egg/patsy/build.py", line 660, in design_matrix_builders
  File "build/bdist.linux-i686/egg/patsy/build.py", line 424, in _examine_factor_types
  File "build/bdist.linux-i686/egg/patsy/eval.py", line 485, in eval
    pwr_eval = ev.evaluators.PowerLevelEvaluator(pwr_mdc, 
  File "build/bdist.linux-i686/egg/patsy/eval.py", line 468, in _eval
    mach_res_dir)
  File "build/bdist.linux-i686/egg/patsy/compat.py", line 117, in call_and_wrap_exc
  File "build/bdist.linux-i686/egg/patsy/eval.py", line 127, in eval
    if type(first_pred) == pan.Series:
  File "<string>", line 1, in <module>
NameError: name 'np' is not defined
```
The original data was transformed using `np.log10(x)`.

**Comment by josef-pkt** [comment]
Ok, in `base.model.from_formula` we save the original dataframe
`        mod.data.frame = data`

(I didn't know)

So the `try except` should be just a precaution and not necessary in standard usage (if `frame` hasn't been deleted)

**Comment by josef-pkt** [comment]
@chrimobs  about `NameError: name 'np' is not defined`
Do you have `np` in the environment where you unpickle? All functions used in the formula need to be available in the "environment". There are still some limitations in this for unpickling that we haven't implemented a work around.

patsy's use of the surrounding environment with `eval` (following R and other weird packages) is the major problem for pickling formulas.

**Comment by chrimobs** [comment]
Yes, `np` is imported in the environment.

**Comment by josef-pkt** [comment]
@chrimobs I can replicate it the `np.log` problem, and I thought that was supposed to work with the evel depth, but it doesn't.
I'm stuck on this, I don't see much with the debugger. And I never tried to understand the **evil** EvalEnvironment.

workaround: add a new column to the data frame with the transformed variable.

Solution for this requires a formula expert, which I'm not.

I will prepare a PR with the fix for original data.

**Comment by josef-pkt** [comment]
PR is #2375

about EvalEnvironment: I didn't try enough depth levels before
`np.` is available with depth=0  which I guess is one of our modules
both `np.log`  and user `log` are available with depth=3

I guess either another `try except` for "depth search" or and option for the user will be necessary.

**Comment by josef-pkt** [comment]
@chrimobs   My examples work now with the two `try except` in my PR, including `log` or `np.log`
You could give it a try with your cases.
I'm out of time right now, but will try to add the unit tests and merge later today.

**Comment by josef-pkt** [comment]
closing this, this part is fixed by #2375, 

refactor for using patsy's pickling support is issue #2469
